framework/src/node/state_machine/state_machine.ts

```diff
@@ -0,0 +1,241 @@
+	VerifyStatus,
+	BlockVerifyContext,
+	VerificationResult,
+	CommandVerifyContext,
+	CommandExecuteContext,
+} from './types';
+
+export interface StateMachineCommand {
+	id: number;
+	schema: Schema;
+	verify?: <T = unknown>(ctx: CommandVerifyContext<T>) => Promise<VerificationResult>;
+	execute: <T = unknown>(ctx: CommandExecuteContext<T>) => Promise<void>;
+}
+
+export interface StateMachineModule {
+	id: number;
+	commands: StateMachineCommand[];
+	verifyTransaction?: (ctx: TransactionVerifyContext) => Promise<VerificationResult>;
+	afterGenesisBlockExecute?: (ctx: GenesisBlockExecuteContext) => Promise<void>;
+	verifyBlock?: (ctx: BlockVerifyContext) => Promise<void>;
+	beforeBlockExecute?: (ctx: BlockExecuteContext) => Promise<void>;
+	afterBlockExecute?: (ctx: BlockExecuteContext) => Promise<void>;
+	beforeTransactionExecute?: (ctx: TransactionExecuteContext) => Promise<void>;
+	afterTransactionExecute?: (ctx: TransactionExecuteContext) => Promise<void>;
+}
+
+export class StateMachine {
+	private readonly _modules: StateMachineModule[] = [];
+	private readonly _systemModules: StateMachineModule[] = [];
+
+	public registerModule(mod: StateMachineModule): void {
+		this._validateExistingModuleID(mod.id);
+		this._modules.push(mod);
+	}
+
+	public registerSystemModule(mod: StateMachineModule): void {
+		this._validateExistingModuleID(mod.id);
+		this._systemModules.push(mod);
+	}
+
+	public async executeGenesisBlock(ctx: GenesisBlockContext): Promise<void> {
+		const blockContext = ctx.createGenesisBlockExecuteContext();
+		for (const mod of this._systemModules) {
+			if (mod.afterGenesisBlockExecute) {
+				await mod.afterGenesisBlockExecute(blockContext);
+			}
+		}
+		for (const mod of this._modules) {
+			if (mod.afterGenesisBlockExecute) {
+				await mod.afterGenesisBlockExecute(blockContext);
+			}
+		}
+	}
+
+	public async verifyTransaction(ctx: TransactionContext): Promise<VerificationResult> {
+		const transactionContext = ctx.createTransactionVerifyContext();
+		try {
+			for (const mod of this._systemModules) {
+				if (mod.verifyTransaction) {
+					const result = await mod.verifyTransaction(transactionContext);
+					if (result.status !== VerifyStatus.OK) {
+						return result;
+					}
+				}
+			}
+			for (const mod of this._modules) {
+				if (mod.verifyTransaction) {
+					const result = await mod.verifyTransaction(transactionContext);
+					if (result.status !== VerifyStatus.OK) {
+						return result;
+					}
+				}
+			}
+			const targetModule = this._findModule(ctx.transaction.moduleID);
+			if (!targetModule) {
+				throw new Error(`Module with ID ${ctx.transaction.moduleID} is not registered.`);
+			}
+			// FIXME: Update assetID to commandID
+			const command = targetModule.commands.find(c => c.id === ctx.transaction.assetID);
+			if (!command) {
+				throw new Error(
+					`Module with ID ${ctx.transaction.moduleID} does not have command with ID ${ctx.transaction.assetID} registered.`,
+				);
+			}
```

Comment:
extract this to a private function `_getCommand`? and use it in other hooks too

Reply:
i was thinking that, but it will be just wrapping one `find` function like below, so it wouldnt make sense
```
private _findCommand(id: number, mod: StateMachineModule): StateMachineCommand | undefined {
		const command = mod.commands.find(c => c.id === id);
		return command;
	}
```

Reply:
I see the `_findCommand` private usage is just to find the command and later used to get the command, so why not just change it to getCommand put the logic there? unless we have `_findCommand` used differently

Reply:
there are 2 places fining the command are used, but both have different logic after finding it